src/h3lib/lib/polyfill.c

```diff
@@ -0,0 +1,422 @@
+    IterCellsPolygonCompact iter = {// Initialize output properties. The first
+                                    // valid cell will be set in iterStep
+                                    .cell = getBaseCell(0),
+                                    .error = E_SUCCESS,
+                                    // Save input arguments
+                                    ._polygon = polygon,
+                                    ._res = res,
+                                    ._flags = flags,
+                                    ._bboxes = NULL,
+                                    ._started = false};
+
+    if (res < 0 || res > MAX_H3_RES) {
+        iterErrorPolygonCompact(&iter, E_RES_DOMAIN);
+        return iter;
+    }
+
+    if (flags != 0) {
+        iterErrorPolygonCompact(&iter, E_OPTION_INVALID);
+        return iter;
+    }
+
+    // Initialize bounding boxes for polygon and any holes. Memory allocated
+    // here must be released through iterDestroyPolygonCompact
+    iter._bboxes = H3_MEMORY(malloc)((polygon->numHoles + 1) * sizeof(BBox));
+    if (!iter._bboxes) {
+        iterErrorPolygonCompact(&iter, E_MEMORY_ALLOC);
+        return iter;
+    }
+    bboxesFromGeoPolygon(polygon, iter._bboxes);
+
+    // Start the iterator by taking the first step.
+    // This is necessary to have a valid value after initialization.
+    iterStepPolygonCompact(&iter);
+
+    return iter;
+}
+
+/**
+ * Increment the polyfill iterator, running the polygon to cells algorithm.
+ *
+ * Briefly, the algorithm checks every cell in the global grid hierarchically,
+ * starting with the base cells. Cells coarser than the target resolution are
+ * checked for complete child inclusion using a bounding box guaranteed to
+ * contain all children.
+ * - If the bounding box is contained by the polygon, output is set to the cell
+ * - If the bounding box intersects, recurse into the first child
+ * - Otherwise, continue with the next cell in sequence
+ *
+ * For cells at the target resolution, a finer-grained check is used according
+ * to the inclusion criteria set in flags.
+ *
+ * @param  iter Iterator to increment
+ */
+void iterStepPolygonCompact(IterCellsPolygonCompact *iter) {
+    H3Index cell = iter->cell;
+    H3Error err;
+
+    // once the cell is H3_NULL, the iterator returns an infinite sequence of
+    // H3_NULL
+    if (cell == H3_NULL) return;
+
+    // For the first step, we need to evaluate the current cell; after that, we
+    // should start with the next cell.
+    if (iter->_started) {
+        cell = nextCell(cell);
+    } else {
+        iter->_started = true;
+    }
+
+    while (cell) {
+        int cellRes = H3_GET_RESOLUTION(cell);
+
+        // Target res: Do a fine-grained check
+        if (cellRes == iter->_res) {
+            // Check if the cell is in the polygon
+            // TODO: Handle other polyfill modes here
+            LatLng center;
+            H3_EXPORT(cellToLatLng)(cell, &center);
```

Comment:
this should have a NEVER check for the error

Reply:
Will add